### DISCUSSION TITLE
KPI Framework

### DISCUSSION THREAD

**Discussion by alesanchezr** [breatheco-de/breatheco-de#2523] [General]
# KPI Framework for the Breathecode.Monitoring app

The best way to monitor business performance is to implement KPI within the breathecode API. A KPI is a metric that measures performance, for example: What is the attendancy rate for cohort 23? What is the drop rate for Acacdemy 2?

Every KPI is calculated in a different way, and its related to a database model entity: bc.models.Academy has KPIs, django.models.User has KPIs, bc.models.Cohort has KPIs, etc.

I'm proposing a new `KPI Script` database model inside the bc.monitoring app that will contain the metadata for each KPI.
The monitoring app will subscribe to any signals from any other bc app and trigger the KPI scripts on demand.
After the script runs it will create_or_replace a `KPI Value` associated to the KPI Script, ModelEntity and ModelEntity.pk

![KPI Scripts](https://user-images.githubusercontent.com/426452/163588400-568a7b82-753e-43d2-918d-72d06ecdfbf8.jpeg)

## Example

Lets say that there is a `KPI Script` associated to bc.admissions.models.Cohort that triggers when a new bc.feedback.models.Answer is received and it sums all the answers for the same cohort and `creates_or_updates` the Total rating for that cohort. This is how the it will look like:

## Creating the script:

A python script must be created similar to the monitoring.scripts, but in this case we can put them in a different folder montoring.kpis. The script must be named with the same name as the KPIScript.slug: `cohort_avg_nps.py`

The KPI Script model will be added like this:

```python
       kpi = KPIScript(
                          slug='cohort_avg_nps',
                          entity_name='breathecode.admission.models.Cohort',
                          min=7,
                          max=10,
                          meassurment_unit="%",
       )
```
## Subscribing to the signal:

n the monioring.receivers we will have to subscribe to the survey_answered and trigger the KPI somehow (as a task), we can us a singleton class, something like `KPITask.execute('cohort_avg_nps', instance)`

## Triggering the signal

When the signal for `survey_answered` is received, the script `monitoring/kpi_script/cohort_avg_nps.py` will be runned and the value of `instance` will be of type Answer and will be automatically given by the signal.

```python
average = Answer.objects.filter(cohort=instance.cohort).aggregate(Avg('score'))

cu = CohortUser.filter(user=answer.user, cohort=answer.cohort).first()
if cu is not None:
       kpi = KPIValue.create_or_update(
                          value=average,
                          entity_id=instance.cohort.id,
                          note="any comments woth mentioning"
       )
```

## Script performance

We need to add performance measurement indicators to the scripts (success or failure), for example:

| KPI                                                  | description                                                                     |
| -------------------------------- | --------------------------------------------------- |
| KPIValue.last_execution_time     | timedelta from start to finish of the execution            |
| KPIValue.last_execution_status  | PENDING, SUCCESS, WARNING or ERROR                 |
| KPIValue.last_execution_details | string with details about the warning or error              |

We can have an email alert if the execution time is to big.

**Comment by jefer94** [comment]
Is possible implement a dispatcher when the signal is trigger, the KPI client should decide if the script should be executed and ends redirecting the flow to Celery (the script run in celery)

In the case KPIValue will be implemented, it should have the fields `id`, `model`, `pki_id`, `key` and `value`

| field | type | value |
|---|---|---|
| id | PK | 1 |
| model | str(4, 20) | feedback.Survey |
| model_id | Unrelated FK | 1 |
| pki_id | FK | 1 |
| key | slug | survey-response-rate |
| value | float(0, 100)| 93 |

Also is possible think about create a dashboard for the logging that works inside admin, it should be possible extending `breathecode/utils/io/logger.py` or saving a report in a bucket, the last system I see of logging have a concept of stakeholder

Is possible KPIScript also have a description of what's do it

The dispatcher should be something like

```py
@shared_task(bind=True, base=BaseTaskWithRetry)
def dispatch_kpi(self, script: str, model: str, pk: Any) -> None:
    """dispatch a pki script"""
    
    ...
    
    KPI.execute(script, model, pk)
```

```py
class KPI:
    def execute(self, script: str, model: str, pk: Any) -> None:
        """execute the PKI script if is possible"""
    
        if PATH_NOT_EXISTS:
            return
    
        ...
        
        # think about send emails or use a extended system of logs
```

```py
    path('measures/<str:script>/<int:fk>', WaitingListView.as_view(), name='measures_script_fk'),  # just one pki
    path('measures/<int:fk>', WaitingListView.as_view(), name='measures_script_fk'),  # many pki
```

**Comment by jefer94** [comment]
The possible issue maybe should occur is the queries become more slower, that should need a redesign of admin for optimize the deliver of data and reduce the number of queries to the db

**Comment by alesanchezr** [comment]
This discussions is closed at this was already implemented using the DataFlow project.
https://github.com/breatheco-de/dataflow-project-template
